src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs

```diff
@@ -23,20 +27,44 @@ internal class VsCodeWindowManager : IVsCodeWindowManager, IVsCodeWindowEvents
             private readonly IVsCodeWindow _codeWindow;
             private readonly ComEventSink _sink;
             private readonly IOptionService _optionService;
+            private readonly IThreadingContext _threadingContext;
+            private readonly WorkspaceRegistration _workspaceRegistration;
 
-            private INavigationBarController _navigationBarController;
-            private IVsDropdownBarClient _dropdownBarClient;
+            private INavigationBarController? _navigationBarController;
+            private IVsDropdownBarClient? _dropdownBarClient;
 
             public VsCodeWindowManager(TLanguageService languageService, IVsCodeWindow codeWindow)
             {
                 _languageService = languageService;
                 _codeWindow = codeWindow;
 
                 var workspace = languageService.Package.ComponentModel.GetService<VisualStudioWorkspace>();
-                _optionService = workspace.Services.GetService<IOptionService>();
+                _optionService = workspace.Services.GetRequiredService<IOptionService>();
+
+                _threadingContext = languageService.Package.ComponentModel.GetService<IThreadingContext>();
 
                 _sink = ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
                 _optionService.OptionChanged += OnOptionChanged;
+
+                ErrorHandler.ThrowOnFailure(_codeWindow.GetBuffer(out var buffer));
+                var textContainer = _languageService.EditorAdaptersFactoryService.GetDataBuffer(buffer).AsTextContainer();
+                _workspaceRegistration = CodeAnalysis.Workspace.GetWorkspaceRegistration(textContainer);
+                _workspaceRegistration.WorkspaceChanged += OnWorkspaceRegistrationChanged;
```

Comment:
Should this subscription happen in AddAdornments? Otherwise if one of these gets created but then never called, we're still leaving this listening?

Reply:
(it's also consistent with the code in RemoveAdornments doing the opposite)